File: go-RayTracer/image.go
package main

import (
	"fmt"
	"io"
)

// Color represents an RGB color

// Image represents a simple image with width, height, and pixel data
type Image struct {
	width, height int
	pixels        []Color
}

// NewImage creates a new image with the specified dimensions
func NewImage(width, height int) *Image {
	return &Image{
		width:  width,
		height: height,
		pixels: make([]Color, width*height),
	}
}

// Pixel returns a reference to the pixel at position (x, y)
func (img *Image) Pixel(x, y int) *Color {
	return &img.pixels[y*img.width+x]
}

func (img *Image) SetPixel(x, y int, color Color) {
	img.pixels[y*img.width+x] = color
}

// WriteToStream outputs the image in PPM format to the provided writer
func (img *Image) WriteToStream(out io.Writer) {
	// Output the PPM header
	fmt.Fprintf(out, "P3\n%d %d\n255\n", img.width, img.height)

	// Output all pixels
	for j := 0; j < img.height; j++ {
		for i := 0; i < img.width; i++ {
			WriteColor(out, img.pixels[j*img.width+i])
		}
	}
}

File: python-Raytracer/image.py
from vec3 import Color
from color import write_color


class Image:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.pixels = [Color(0.0, 0.0, 0.0) for _ in range(width * height)]

    def set_pixel(self, x, y, color):
        self.pixels[y * self.width + x] = color

    def write_to(self, out):
        """Write the image to the given file stream in PPM format"""
        # Write PPM header
        out.write(f"P3\n{self.width} {self.height}\n255\n")

        # Write all pixels
        for j in range(self.height):
            for i in range(self.width):
                write_color(out, self.pixels[j * self.width + i])

        return True
